lib/diff_darwin.js

```diff
@@ -0,0 +1,1332 @@
+                    Array.prototype.push.apply(args, subargs);
+                }
+                return predicate;
+            });
+            args.unshift(subfilters.join(' OR '));
+            return args;
+        }
+        case 'not': {
+            let args = [];
+            let subfilters = _.map(filter.slice(1), function (filter) {
+                let subargs = filterToPredicate(filter);
+                let predicate = subargs[0];
+                if (['all', 'any', 'none'].indexOf(filter[0]) !== -1) {
+                    predicate = '(' + predicate + ')';
+                }
+                subargs = subargs.slice(1);
+                if (subargs[0]) {
+                    Array.prototype.push.apply(args, subargs);
+                }
+                return predicate;
+            });
+            if (subfilters.length > 1) {
+                args.unshift(`NOT (${subfilters.join(' OR ')})`);
+            } else {
+                args.unshift(`NOT ${subfilters}`);
+            }
+            return args;
+        }
+        default:
+            throw `Filter operator ${op} not implemented.`;
+    }
+}
+
+let operations = {
+    addLayer: function (layer, beforeLayerId, os) {
+        let layerVar = _.camelCase(layer.id) + 'Layer';
+        let layerInitArgs = {
+            initWithIdentifier: Literal(layer.id),
+        };
+        let beforeLayerVar = _.camelCase(beforeLayerId) + 'Layer';
+        let statements = [];
+        let layerClass;
+        let layoutStatements = _.map(layer.layout, function (jsValue, jsPropName) {
+            if (!layerClass) {
+                layerClass = getLayerClassByProperty(jsPropName);
+            }
+            let propName = getLayerPropertyName(jsPropName, os);
+            let value = valueToStyleValue(jsValue, jsPropName, os);
+            return Assignment(KeyPath([layerVar, propName]), value);
+        });
+        if (layoutStatements.length) {
+            statements.push(WithStatement(layerVar, layoutStatements));
+        }
+        let paintStatements = _.map(layer.paint, function (jsValue, jsPropName) {
+            if (!layerClass) {
+                layerClass = getLayerClassByProperty(jsPropName);
+            }
+            let propName = getLayerPropertyName(jsPropName, os);
+            let value = valueToStyleValue(jsValue, jsPropName, os);
+            return Assignment(KeyPath([layerVar, propName]), value);
+        });
+        if (paintStatements.length) {
+            statements.push(WithStatement(layerVar, paintStatements));
+        }
+        if (!layerClass) {
+            layerClass = `MGL${upperCamelCase(layer.type)}StyleLayer`;
+        }
+        let sourceId;
+        let sourceVar;
+        if ('source' in layer) {
+            sourceId = layer.source;
+            sourceVar = _.camelCase(layer.source) + 'Source';
+            layerInitArgs.source = sourceVar;
+        }
+        let refLayerId;
+        let refLayerVar;
+        if ('ref' in layer) {
```

Comment:
Style diffs with refs are as mess. In `lib/diff.js` I ignored that refs will ever exist. On the plus side, refs are going away anyway.